src/migrator/utils.ts

```diff
@@ -23,12 +24,137 @@ export type CategorizedMigrations = {
 type SbvrUtils = typeof sbvrUtils;
 export type MigrationTuple = [string, Migration];
 export type MigrationFn = (tx: Tx, sbvrUtils: SbvrUtils) => Resolvable<void>;
-export type Migration = string | MigrationFn;
 export type RunnableMigrations = { [key: string]: Migration };
+export type RunnableAsyncMigrations = { [key: string]: AsyncMigration };
 export type Migrations = CategorizedMigrations | RunnableMigrations;
+export type AsyncMigrationFn = (
+	tx: Tx,
+	options: { batchSize: number },
+	sbvrUtils: SbvrUtils,
+) => Resolvable<Result>;
+
+type AddFn<T extends {}, x extends 'sync' | 'async'> = T & {
+	syncFn: MigrationFn;
+	asyncFn: AsyncMigrationFn;
+} & {
+	[key in `${x}Sql`]?: undefined;
+};
+type AddSql<T extends {}, x extends 'sync' | 'async'> = T & {
+	[key in `${x}Fn`]?: undefined;
+} & {
+	[key in `${x}Sql`]: string;
+};
+
+type BaseAsyncMigration = {
+	type?: 'async';
+	delayMS?: number | undefined;
+	backoffDelayMS?: number | undefined;
+	errorThreshold?: number | undefined;
+	asyncBatchSize?: number | undefined;
+	finalize?: boolean | undefined;
+};
+export type AsyncMigration =
+	| AddFn<BaseAsyncMigration, 'async' | 'sync'>
+	| AddSql<BaseAsyncMigration, 'async' | 'sync'>
+	| AddFn<AddSql<BaseAsyncMigration, 'async'>, 'sync'>
+	| AddFn<AddSql<BaseAsyncMigration, 'sync'>, 'async'>;
 
+export function isAsyncMigration(
+	migration: AsyncMigration | RunnableMigrations,
+): migration is AsyncMigration {
+	return (migration as AsyncMigration).type === MigrationCategories.async;
+}
+
+export function areCategorizedMigrations(
+	migrations: Migrations,
+): migrations is CategorizedMigrations {
+	return Object.keys(migrations).some((key) => key in MigrationCategories);
```

Comment:
I think this should be
```suggestion
	return Object.keys(migrations).every((key) => key in MigrationCategories);
```
?
Or maybe something like
```suggestion
	const keys = Object.keys(migrations);
	const containsCategories = keys.some((key) => key in MigrationCategories);
	if (containsCategories && !keys.every((key) => key in MigrationCategories)) {
		throw new Error('Mixing categorized and uncategorized migrations is not supported');
	}
	return containsCategories;
```
to ensure that people don't mix/match migrations (which will presumably be an accident and break things?)

Reply:
I'm going for the mixing category protection. Thanks